File: problems/1971_find_if_path_exists_in_graph.py
class UnionFind:
    size: int
    _roots: list[int]
    _ranks: list[int]

    def __init__(self, size: int):
        if size <= 0:
            raise Exception("Invalid size.")
        self.size = size
        self._roots = [i for i in range(size)]
        self._ranks = [1] * size

    def union(self, a: int, b: int) -> None:
        root_a = self.find_root(a)
        root_b = self.find_root(b)
        rank_a = self._ranks[root_a]
        rank_b = self._ranks[root_b]
        if rank_a < rank_b:
            self._roots[root_a] = root_b
        elif rank_a > rank_b:
            self._roots[root_b] = root_a
        else:
            self._roots[root_b] = root_a
            self._ranks[root_a] += 1

    def connected(self, a: int, b: int) -> bool:
        root_a = self.find_root(a)
        root_b = self.find_root(b)
        return root_a == root_b

    def find_root(self, index: int) -> int:
        if self._roots[index] == index:
            return index
        parent = self._roots[index]
        root = self.find_root(parent)
        self._roots[index] = root
        return root


class Solution:
    def validPath(
        self, n: int, edges: list[list[int]], source: int, destination: int
    ) -> bool:
        uf = UnionFind(n)
        for [u, v] in edges:
            uf.union(u, v)
        return uf.connected(source, destination)

File: problems/1971-find-if-path-exists-in-graph.ts
export {}

class UnionFind {
  readonly size: number
  private roots: number[]
  private ranks: number[]

  constructor(size: number) {
    if (size <= 0) {
      throw new Error("Invalid size.")
    }
    this.size = size
    this.roots = Array.from({ length: size }, (_, k) => k)
    this.ranks = Array(size).fill(1)
  }

  union(a: number, b: number): void {
    const { size, roots, ranks } = this
    if (a < 0 || a >= size || b < 0 || b >= size) {
      throw new Error("Invalid input. `a` and/or `b` out of range.")
    }
    const rootA = this.findRoot(a)
    const rootB = this.findRoot(b)
    const rankA = ranks[rootA]
    const rankB = ranks[rootB]
    if (rankA < rankB) {
      roots[rootA] = rootB
    } else if (rankA > rankB) {
      roots[rootB] = rootA
    } else {
      roots[rootB] = rootA
      ranks[rootA] += 1
    }
  }

  connected(a: number, b: number): boolean {
    const { size } = this
    if (a < 0 || a >= size || b < 0 || b >= size) {
      throw new Error("Invalid input. `a` and/or `b` out of range.")
    }
    const rootA = this.findRoot(a)
    const rootB = this.findRoot(b)
    return rootA === rootB
  }

  findRoot(index: number): number {
    const { roots } = this
    if (index === roots[index]) {
      return index
    }
    const parent = roots[index]
    const newRoot = this.findRoot(parent)
    roots[index] = newRoot
    return newRoot
  }
}

function validPath(
  n: number,
  edges: number[][],
  source: number,
  destination: number
): boolean {
  const uf = new UnionFind(n)
  for (const [u, v] of edges) {
    uf.union(u, v)
  }
  return uf.connected(source, destination)
}
